### ISSUE TITLE
Evaluation of Precision@k and Recall@k Results 

### ISSUE THREAD

**Issue by Shadz13** [massquantity/LibRecommender#45]
I have split the data using the random split algorithm to obtain a train and test set. The test set is then split to obtain a validation set (i.e a dataset that is unseen by the model).

The code below shows how the data is split:
```
      train_data_main, test_val_data= random_split(data, test_size=0.2) 
      test_data = test_val_data.sample(frac = 0.5) 
      val_data = test_val_data.drop(test_data_main_random_split.index) 
```

NB: for this example a dummy train data (13k) that is much smaller therefore the auc results will be lower.

I have obtained the precision@k, recall@k, map@k and ndcg@k results for the test_data and val_data. For further testing I also used the training set as a test set to obtain precision@k, recall@k, map@k and ndcg@k results.
The image below shows the impact of varying the ratio of relevant items in a four out of ten recommendation list:
![PkRk](https://user-images.githubusercontent.com/42578613/109622755-204fe980-7b45-11eb-8314-2aaea7aaa8fc.PNG)

For the model I generated, I have k relevant items of 3 and selected out of 3 recommendation, by setting k in the evaluation script to 3. (assumption made that setting k in the evaluation script means 3 rec items)
```
eval_rec: 100%|██████████| 822/822 [00:01<00:00, 746.90it/s]
	 eval log_loss: 0.6839
	 eval log_loss: 0.6839
	 eval balanced accuracy: 0.5985
	 eval roc_auc: 0.6773
	 eval pr_auc: 0.6530
	 eval precision@3: 0.1383
	 eval recall@3: 0.3713
	 eval map@3: 0.2307
	 eval ndcg@3: 0.2750
==============================
train: 100%|██████████| 209/209 [00:01<00:00, 106.02it/s]
eval_pred: 100%|██████████| 1/1 [00:00<00:00, 120.27it/s]
eval_rec:  11%|█▏        | 93/822 [00:00<00:00, 916.23it/s]Epoch 2 elapsed: 1.996s
	 train_loss: 0.6909
eval_rec: 100%|██████████| 822/822 [00:01<00:00, 433.53it/s]
	 eval log_loss: 0.6433
	 eval log_loss: 0.6433
	 eval balanced accuracy: 0.6849
	 eval roc_auc: 0.7347
	 eval pr_auc: 0.6992
	 eval precision@3: 0.1590
	 eval recall@3: 0.4333
	 eval map@3: 0.2691
	 eval ndcg@3: 0.3195
```

How is it possible that both precision@3 increases while recall@3 increases, unless it not evaluating within 3 recommended items but the total number of items altogether.

I outputted the precision @k  and recall @k  for the validation set where recommending items is 3.Code shown below.
 Even though recommended items are 3 can the results from val_df be benchmarked directly with the above results. 
```
##### Obtain/Store recommended list, associated items and user
     val_rec, val_rec_user, val_rec_item = [], [] , []
     for user_ in  val_data_main['user']:
          val_rec.append(list(itertools.chain(*deepFM_best.recommend_user(user=user_, n_rec=3))))    
          val_rec_user.append(user_)    
          val_rec_item.append(list(val_data_main[val_data_main['user']==user_].reset_index()['Benefit_Cover_Type'])) 

 ##### Convert to a dataframe and rearrange columns so that the probabilities are after the recommendations        
     val_rec = pd.DataFrame(val_rec).reindex([0,2,4,1,3,5], axis=1) 
     val_rec.columns = ['rec_1', 'rec_2','rec_3','prob_1','prob_2','prob_3']
     val_rec_user = pd.DataFrame(val_rec_user)
     val_rec_item = pd.DataFrame(val_rec_item)

##### Rename item columns
     t_list=[]
     for i in range(1,len(val_rec_item .columns)+1):
              t_list.append('item_'+str(i))
     val_rec_item.columns = t_list  
     val_rec_user.columns = ['user']
##### Merge user dataframe with dataframe of items (used for recommendation list)
     val_rec_results = val_rec_user.merge(val_rec_item, left_index=True, right_index=True, how='inner') 

###### I- item information 
     I=val_rec_results.drop(columns=['user'])
     for ii in t_list:
         I[ii]=val_rec_results[ii].map(c)

##### Obtain y_true_lst 
     y_true_list = {k:list(v) for k,v in zip(val_rec_results['user'],I.values)}

 #### drop all items and prob values so that R has just recommendations    
     val_rec_results = val_rec_results.merge(val_rec, left_index=True, right_index=True, how='inner')
     R1 = val_rec_results
     for ii in t_list:
         R1 = R1.drop(columns=[ii])
     R = R1.drop(columns=['user','prob_1','prob_2','prob_3'])
     
##### Obtain y_reco_lst     
     y_reco_list = {k:list(v) for k,v in zip(val_rec_results['user'], R.values)}
   
##### map to convert cat codes to original labels            
     val_rec_results['rec_1']=val_rec_results['rec_1'].map(d)
     val_rec_results['rec_2']=val_rec_results['rec_2'].map(d)
     val_rec_results['rec_3']=val_rec_results['rec_3'].map(d)  
      
     val_df = pd.DataFrame()
     val_df['precision_all_@'+str(k)] = [round(precision_at_k(y_true_list, y_reco_list,list(val_rec_results['user']), k=3),4)]
     val_df['recall_all_@'+str(k)]    = [round(recall_at_k(y_true_list, y_reco_list,list(val_rec_results['user']), k=3),4)]
     val_df['map_all_@'+str(k)]       = [round(map_at_k(y_true_list, y_reco_list,list(val_rec_results['user']), k=3),4)]
     val_df['ndcg_all_@'+str(k)]      = [round(ndcg_at_k(y_true_list, y_reco_list,list(val_rec_results['user']), k=3),4)]

```

**Comment by massquantity** [comment]
Why isn't it possible? As the model is training, the recommended results become more accurate, then precision and recall can increase at the same time.
